pkg/datapath/neighbor/desired_neighbor_calculator.go

```diff
@@ -0,0 +1,305 @@
+				// New object, but no modifications
+				return &DesiredNeighbor{
+					DesiredNeighborKey: old.DesiredNeighborKey,
+					Status:             old.Status,
+				}
+			})
+			if err != nil {
+				errs = errors.Join(errs, fmt.Errorf("failed to insert desired neighbor: %w", err))
+			}
+		}
+		if errs != nil {
+			health.Degraded("desired neighbor calculator errored", errs)
+		}
+
+		// If the forwardable IP table is initialized in the read transaction,
+		// then mark the desired neighbor table as initialized as well.
+		if init, _ := c.ForwardableIPTable.Initialized(rx); init {
+			c.desiredNeighborInitialized(tx)
+		}
+
+		tx.Commit()
+
+		if needFullSyn {
+			needFullSyn = false
+			resyncTimer.Reset(resyncInterval)
+		}
+
+		select {
+		case <-ctx.Done():
+			// If we are shutting down, stop the loop
+			return nil
+
+		case <-fipWatch:
+			// If there are updates to the forwardable IPs, do a partial sync
+			continue
+
+		case <-l2DevWatch:
+			// If there are updates to the L2 devices, do a full sync
+			needFullSyn = true
+
+		case <-routeWatch:
+			// If there are updates to the routes, do a full sync
+			needFullSyn = true
+
+		case <-resyncTimer.C:
+			// If the resync timer fires, do a full sync
+			needFullSyn = true
+		}
+	}
+}
+
+var errNodeIPNotRoutable = errors.New("remote node IP is non-routable")
+
+func (c *desiredNeighborCalculator) getNextHopIP(fip netip.Addr, ifindex int) (nextHopIP netip.Addr, err error) {
+	c.Metrics.NexthopLookupCount.Inc()
+
+	// Figure out whether nodeIP is directly reachable (i.e. in the same L2)
+	routes, err := c.FuncsGetter.Get().RouteGetWithOptions(fip.AsSlice(), &netlink.RouteGetOptions{
+		OifIndex: ifindex,
+		FIBMatch: true,
+	})
+	if err != nil && !errors.Is(err, unix.EHOSTUNREACH) && !errors.Is(err, unix.ENETUNREACH) {
+		return netip.Addr{}, fmt.Errorf("failed to retrieve route for remote node IP: %w", err)
+	}
+	if len(routes) == 0 {
+		return netip.Addr{}, errNodeIPNotRoutable
+	}
+
+	nextHopIP = fip
+	for _, route := range routes {
+		if route.Gw != nil {
+			// nodeIP is in a different L2 subnet, so it must be reachable through
+			// a gateway. Perform neighbor discovery to the gw IP addr instead of
+			// nodeIP. NOTE: We currently don't handle multipath, so only one gw
+			// can be used.
+
+			if route.Gw.To4() == nil {
+				nextHopIP, _ = netip.AddrFromSlice(route.Gw.To16())
+			} else {
+				nextHopIP, _ = netip.AddrFromSlice(route.Gw.To4())
+			}
+			break
+		}
+
+		// Select a gw for the specified link if there are multi paths to the nodeIP
+		// For example, the nextHop to the nodeIP 9.9.9.9 from eth0 is 10.0.1.2,
+		// from eth1 is 10.0.2.2 as shown bellow.
+		//
+		// 9.9.9.9 proto bird metric 32
+		//        nexthop via 10.0.1.2 dev eth0 weight 1
+		//        nexthop via 10.0.2.2 dev eth1 weight 1
+		//
+		// NOTE: We currently don't handle multiple next hops, so only one next hop
+		// per device can be used.
```

Comment:
would probably be good to document in a GH issue what would roughly be needed in order to support this in future so we don't forget about it once we pick that up